### PR TITLE
[wasm] Enable more networking tests

### PR DESCRIPTION
**Pull request by pavelsavara**
- System.Net.Http\tests\FunctionalTests is now included in tests
- I moved System.Net.Http.Functional.Tests and System.Net.WebSockets.Client.Tests to run in chrome instead of V8 via `WasmTestOnBrowser`
- I used `[ActiveIssue]` https://github.com/dotnet/runtime/issues/37669 for missing crypto MD5
- I used `[ActiveIssue]` https://github.com/dotnet/runtime/issues/42852 for loopback support
- I used `[ActiveIssue]` https://github.com/dotnet/runtime/issues/53591 - GET/HEAD method cannot have body
- I used `[ActiveIssue]` https://github.com/dotnet/runtime/issues/53592 - 'TRACE' HTTP method is unsupported.
- I used `[ActiveIssue]` https://github.com/dotnet/runtime/issues/53018 - Failed to fetch
- I used `[ActiveIssue]` https://github.com/dotnet/runtime/issues/53281 for WebSocket problems on Mono/Windows
- I used `[SkipOnPlatform]` when I was confident that we do not plan to support the feature
- Fixed `IsMsQuicSupported` to return bool instead of throwing

### CODE REVIEW

File: src/libraries/System.Net.Http/tests/FunctionalTests/System.Net.Http.Functional.Tests.csproj [hunk 1]
```diff
@@ -6,6 +6,7 @@
     <DefineConstants>$(DefineConstants);SYSNETHTTP_NO_OPENSSL;HTTP3</DefineConstants>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
     <IncludeRemoteExecutor>true</IncludeRemoteExecutor>
+    <Scenario>WasmTestOnBrowser</Scenario>
```
Comment (akoeplinger) [inline src/libraries/System.Net.Http/tests/FunctionalTests/System.Net.Http.Functional.Tests.csproj:9]: I think we should have a condition here so it only applies to scenario to browser:
```suggestion
    <Scenario Condition="'$(TargetsBrowser)'=='true'">WasmTestOnBrowser</Scenario>
```

File: src/libraries/System.Net.Http/tests/FunctionalTests/HttpContentTest.cs [hunk 2]
```diff
@@ -829,6 +836,7 @@ public async Task ReadAsStreamAsync_Buffered_IgnoresCancellationToken(bool readS
         [InlineData(true)]
         [InlineData(false)]
         [ActiveIssue("https://github.com/dotnet/runtime/issues/39187", TestPlatforms.Browser)]
+        [ActiveIssue("https://github.com/dotnet/runtime/issues/42852", TestPlatforms.Browser)]
```
Comment (akoeplinger) [inline src/libraries/System.Net.Http/tests/FunctionalTests/HttpContentTest.cs:839]: #39187 seems to be a different issue? we should replace it with the one you added here

Comment (pavelsavara) [reply src/libraries/System.Net.Http/tests/FunctionalTests/HttpContentTest.cs:839]: Both of them possibly apply still.  I didn't try to resolve #39187.
There are at least 7 other places with 39187 on it.
I prefer to keep 39187 open until the whole test passes.

File: src/libraries/System.Net.Http/tests/FunctionalTests/System.Net.Http.Functional.Tests.csproj (hunk 1 above)
Comment (pavelsavara) [reply src/libraries/System.Net.Http/tests/FunctionalTests/System.Net.Http.Functional.Tests.csproj]: done

File: src/libraries/Common/tests/System/Net/Http/HttpClientHandlerTest.MaxConnectionsPerServer.cs [hunk 3]
```diff
@@ -64,6 +64,7 @@ public void Set_ValidValues_Success(int validValue)
         [InlineData(3, 2, true)]
         [InlineData(3, 5, false)]
         [OuterLoop("Uses external servers")]
+        [SkipOnPlatform(TestPlatforms.Browser, "MaxConnectionsPerServer not supported on Browser")]
```
Comment (mdh1418) [inline src/libraries/Common/tests/System/Net/Http/HttpClientHandlerTest.MaxConnectionsPerServer.cs:67]: Should `SkipOnPlatform` be added for the rest of the methods that use `MaxConnectionsPerServer`? Or are we only adding it to this method because it is part of the OuterLoop?

We should address the other tests in a followup PR if not in this one.

Comment (pavelsavara) [reply src/libraries/Common/tests/System/Net/Http/HttpClientHandlerTest.MaxConnectionsPerServer.cs]: The whole class should have it actually and this is a duplicate, thanks for catching it.

File: src/libraries/Common/tests/System/Net/Http/HttpClientHandlerTest.RemoteServer.cs [hunk 4]
```diff
@@ -843,6 +869,7 @@ public async Task SendAsync_SendSameRequestMultipleTimesDirectlyOnHandler_Succes
         [OuterLoop("Uses external server")]
         [Theory]
         [MemberData(nameof(Http2Servers))]
+        [ActiveIssue("https://github.com/dotnet/runtime/issues/53018", TestPlatforms.Browser)]
```
Comment (mdh1418) [inline src/libraries/Common/tests/System/Net/Http/HttpClientHandlerTest.RemoteServer.cs:872]: Does `SendAsync_RequestVersion20_ResponseVersion20` work? Is there something particular about `SendAsync_RequestVersion20_ResponseVersion20IfHttp2Supported` that fails that is not a part of `SendAsync_RequestVersion20_ResponseVersion20`

File: src/libraries/Common/tests/System/Net/Http/HttpClientHandlerTest.RemoteServer.cs [hunk 5]
```diff
@@ -94,6 +96,7 @@ public async Task SendAsync_SimpleGet_Success(Configuration.Http.RemoteServer re
 
         [OuterLoop("Uses external server")]
```
Comment (mdh1418) [inline src/libraries/Common/tests/System/Net/Http/HttpClientHandlerTest.RemoteServer.cs:97]: Should we change these to `[OuterLoop("Uses external servers")]` as well, like in `src/libraries/Common/tests/System/Net/Http/PostScenarioTest.cs`

File: src/libraries/sendtohelixhelp.proj [hunk 6]
```diff
@@ -371,8 +372,8 @@
...
-    <Message Condition="'$(Scenario)' == '' and ('$(TargetOS)' == 'Android' or '$(TargetOS)' == 'iOS' or '$(TargetOS)' == 'iOSSimulator' or '$(TargetOS)' == 'tvOS' or '$(TargetOS)' == 'tvOSSimulator' or '$(TargetOS)' == 'MacCatalyst')" Importance="High" Text="Done building Helix work items. Work item count: @(XHarnessAppBundleToTest->Count())" />
-    <Message Condition="'$(Scenario)' == '' and '$(TargetOS)' != 'Android' and '$(TargetOS)' != 'iOS' and '$(TargetOS)' != 'iOSSimulator' and '$(TargetOS)' != 'tvOS' and '$(TargetOS)' != 'tvOSSimulator' and '$(TargetOS)' != 'MacCatalyst'" Importance="High" Text="Done building Helix work items. Work item count: @(_WorkItem->Count())" />
+    <Message Condition="'$(Scenario)' == '' and ('$(TargetOS)' == 'Android' or '$(TargetOS)' == 'iOS' or '$(TargetOS)' == 'iOSSimulator' or '$(TargetOS)' == 'tvOS' or '$(TargetOS)' == 'tvOSSimulator' or '$(TargetOS)' == 'MacCatalyst')" Importance="High" Text="Done building Helix work items for scenario $(Scenario). Work item count: @(XHarnessAppBundleToTest->Count())" />
```
Comment (mdh1418) [inline src/libraries/sendtohelixhelp.proj:375]: wouldn't the text evaluate to ` for scenario .` if the message condition is `'$(Scenario)' == '' and ...`

File: src/libraries/Common/tests/System/Net/Http/HttpClientHandlerTest.RemoteServer.cs (hunk 5 above)
Comment (pavelsavara) [reply src/libraries/Common/tests/System/Net/Http/HttpClientHandlerTest.RemoteServer.cs:97]: unified everywhere

**Review by mdh1418 — APPROVED** [review]
Thanks, overall it looks good to me!

File: src/libraries/Common/tests/System/Net/Http/HttpClientHandlerTest.RemoteServer.cs (hunk 4 above)
Comment (pavelsavara) [reply src/libraries/Common/tests/System/Net/Http/HttpClientHandlerTest.RemoteServer.cs:872]: some of the HTTP/2 is probably broken, but I want to investigate that under #53018 
The `SendAsync_RequestVersion20_ResponseVersion20` is Windows only test

**Review by radekdoulik — APPROVED** [review]

File: src/libraries/Common/tests/System/Net/Http/PostScenarioTest.cs [hunk 7]
```diff
@@ -211,6 +242,12 @@ public async Task PostNonRewindableContentUsingAuth_PreAuthenticate_Success(Conf
         [Theory, MemberData(nameof(RemoteServersMemberData))]
         public async Task PostAsync_EmptyContent_ContentTypeHeaderNotSent(Configuration.Http.RemoteServer remoteServer)
         {
+            if (remoteServer.HttpVersion.Major >= 2 && PlatformDetection.IsBrowser)
```
Comment (radical) [inline src/libraries/Common/tests/System/Net/Http/PostScenarioTest.cs:245]: I think a different issue should be opened for it, if it is not the same failure as others in #53018, with details of the failure.

File: src/libraries/Common/tests/System/Net/Http/PostScenarioTest.cs [hunk 8]
```diff
@@ -80,11 +95,18 @@ public async Task PostEmptyContentUsingChunkedEncoding_Success(Configuration.Htt
...
+                // [ActiveIssue("https://github.com/dotnet/runtime/issues/53018", TestPlatforms.Browser)]
+                return;
+            }
+
             await PostHelper(remoteServer, string.Empty, new StringContent(string.Empty),
                 useContentLengthUpload: true, useChunkedEncodingUpload: true);
         }
 
         [OuterLoop("Uses external servers")]
+        [SkipOnPlatform(TestPlatforms.Browser, "Not supported on Browser")]
```
Comment (radical) [inline src/libraries/Common/tests/System/Net/Http/PostScenarioTest.cs:109]: What isn't supported here? And same question for the other instances of `"Not supported on Browser"`.

File: src/libraries/Common/tests/System/Net/Http/HttpClientHandlerTest.ServerCertificates.cs [hunk 9]
```diff
@@ -235,7 +235,7 @@ public async Task NoCallback_RevokedCertificate_NoRevocationChecking_Succeeds()
             }
         }
 
-        [OuterLoop("Uses external server")]
+        [OuterLoop("Uses external servers")]
```
Comment (radical) [inline src/libraries/Common/tests/System/Net/Http/HttpClientHandlerTest.ServerCertificates.cs:238]: Do we really need this change? It makes the diff really hard to read.

File: eng/pipelines/libraries/outerloop-mono.yml [hunk 10]
```diff
@@ -38,6 +38,9 @@ jobs:
         # extra steps, run tests
         extraStepsTemplate: /eng/pipelines/libraries/helix.yml
         extraStepsParameters:
+          scenarios:
+          - wasmtestonbrowser
+          - normal
```
Comment (radical) [inline eng/pipelines/libraries/outerloop-mono.yml:43]: do we need both browser, and v8 run?

**Review by radical — CHANGES_REQUESTED** [review]
I would suggest splitting this PR into two:

1. Enable as many of the networking tests that you are enabling here, *without* the *browseronly* part being needed
2. Add the browseronly bit, and enable the tests that can be enabled because of that

**Comment by pavelsavara** [comment]
> I would suggest splitting this PR into two:
> 
> 1. Enable as many of the networking tests that you are enabling here, _without_ the _browseronly_ part being needed
> 2. Add the browseronly bit, and enable the tests that can be enabled because of that

There is not much networking in V8 alone.

**Comment by radical** [comment]
> > I would suggest splitting this PR into two:
> > 
> > 1. Enable as many of the networking tests that you are enabling here, _without_ the _browseronly_ part being needed
> > 2. Add the browseronly bit, and enable the tests that can be enabled because of that
> 
> There is not much networking in V8 alone.

Ok, so none of the new tests being enabled now should run with v8? I'm trying to separate this into:

1. enable the 3 test suites, for the regular v8 run, disable any that don't make sense on v8, or don't work
   - is this zero?

2. Add browseronly bit, enable/disable any tests affected by this. Then we can clean this part up too.

File: eng/pipelines/libraries/outerloop-mono.yml (hunk 10 above)
Comment (pavelsavara) [reply eng/pipelines/libraries/outerloop-mono.yml:43]: I think we use v8 for most tests currently because it's lightweight. 
On the other hand, it doesn't have some features we use. This PR is mostly about networking and that's feature of browser, not feature of js engine. NodeJs has separate implementation AFAIK.

**Comment by pavelsavara** [comment]
> > > I would suggest splitting this PR into two:
> > > 
> > > 1. Enable as many of the networking tests that you are enabling here, _without_ the _browseronly_ part being needed
> > > 2. Add the browseronly bit, and enable the tests that can be enabled because of that
> > 
> > 
> > There is not much networking in V8 alone.
> 
> Ok, so none of the new tests being enabled now should run with v8? I'm trying to separate this into:
> 
> 1. enable the 3 test suites, for the regular v8 run, disable any that don't make sense on v8, or don't work
>    
>    * is this zero?
> 2. Add browseronly bit, enable/disable any tests affected by this. Then we can clean this part up too.

We have ill-named `IsBrowserDomSupported`. We could start with that.

File: eng/pipelines/libraries/outerloop-mono.yml (hunk 10 above)
Comment (radical) [reply eng/pipelines/libraries/outerloop-mono.yml:43]: Then can we use just browser here? IIUC, `wasmtestonbrowser` will run tests run with `normal`(v8)+the networking ones. Not objecting to it, just making sure that this is the intention.

**Comment by pavelsavara** [comment]
- for `System.Net.WebSockets.Client.Tests` it's Condition not met: `WebSocketsSupported` in 100% cases
- for `Common.Tests` you were right, we could keep it in `normal` scenario.
- for `System.Net.WebSockets.Client.Tests` I marked it as `WasmTestOnBrowser` which should run it on browser when available
- I created https://github.com/dotnet/runtime/issues/53591 and https://github.com/dotnet/runtime/issues/53592 with more specific error details for the tests which have specific failure

**Comment by pavelsavara** [comment]
>    * is this zero?

For `System.Net.WebSockets.Client.Tests` yes. 
For `System.Net.Http.Functional.Tests` V8 need to skip the same 78 tests as browser, but for different reason. I made that distinction explicit.

File: src/libraries/Common/tests/System/Net/Http/PostScenarioTest.cs (hunk 8 above)
Comment (pavelsavara) [reply src/libraries/Common/tests/System/Net/Http/PostScenarioTest.cs]: Actually, we don't know yet. all we get is `Failed to fetch` error, which I would like to address in https://github.com/dotnet/runtime/issues/53018
So I changed this to `ActiveIssue` rather.

File: src/libraries/Common/tests/System/Net/Http/PostScenarioTest.cs (hunk 7 above)
Comment (pavelsavara) [reply src/libraries/Common/tests/System/Net/Http/PostScenarioTest.cs]: I created https://github.com/dotnet/runtime/issues/53591 and https://github.com/dotnet/runtime/issues/53592, for cases which have more specific error message and tagged the relevant tests.

File: src/libraries/Common/tests/System/Net/Http/HttpClientHandlerTest.Decompression.cs [hunk 11]
```diff
@@ -170,6 +170,7 @@ public static IEnumerable<object[]> DecompressedResponse_MethodNotSpecified_Orig
         }
 
         [OuterLoop("Uses external servers")]
+        [SkipOnPlatform(TestPlatforms.Browser, "AutomaticDecompression not supported on Browser")]
```
[the remark this answers is not in the document]
Comment (pavelsavara) [reply src/libraries/Common/tests/System/Net/Http/HttpClientHandlerTest.Decompression.cs:173]: @lewing ?

File: src/libraries/Common/tests/System/Net/Http/HttpClientHandlerTest.ServerCertificates.cs (hunk 9 above)
Comment (pavelsavara) [reply src/libraries/Common/tests/System/Net/Http/HttpClientHandlerTest.ServerCertificates.cs:238]: It was suggested by Mitchell we should unify it and I agree. I separated it into different commit on this PR to make it bit easier to review.

File: src/libraries/Common/tests/System/Net/Http/HttpClientHandlerTest.Decompression.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (lewing) [reply src/libraries/Common/tests/System/Net/Http/HttpClientHandlerTest.Decompression.cs:173]: I don't think we have any control over it at this level.

File: src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTestBase.SocketsHttpHandler.cs [hunk 12]
```diff
@@ -16,7 +16,7 @@ public abstract partial class HttpClientHandlerTestBase : FileCleanupTestBase
 
         protected virtual QuicImplementationProvider UseQuicImplementationProvider => null;
 
-        public static bool IsMsQuicSupported => QuicImplementationProviders.MsQuic.IsSupported;
+        public static bool IsMsQuicSupported => PlatformDetection.IsNotBrowser && QuicImplementationProviders.MsQuic.IsSupported;
```
Comment (radical) [inline src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTestBase.SocketsHttpHandler.cs:19]: Should `QuicImplementationProviders.MsQuic.IsSupported` be updated to handle the browser case? That way, if an app tries to use that API, then it would return `false` for them?

Comment (pavelsavara) [reply src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTestBase.SocketsHttpHandler.cs:19]: The problem is that QuicImplementationProviders throws PlatformNotSupportedException on AssemblyInfo level.

Comment (radical) [reply src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTestBase.SocketsHttpHandler.cs:19]: What do you mean by "on AssemblyInfo level"?

File: src/libraries/Common/tests/System/Net/Http/PostScenarioTest.cs [hunk 13]
```diff
@@ -62,30 +65,52 @@ public async Task PostNoContentUsingContentLengthSemantics_Success(Configuration
...
         [OuterLoop("Uses external servers")]
         [Theory, MemberData(nameof(RemoteServersMemberData))]
+        [ActiveIssue("https://github.com/dotnet/runtime/issues/37669", TestPlatforms.Browser)]
         public async Task PostEmptyContentUsingChunkedEncoding_Success(Configuration.Http.RemoteServer remoteServer)
         {
+            if (remoteServer.HttpVersion.Major >= 2 && PlatformDetection.IsBrowser)
+            {
+                // [ActiveIssue("https://github.com/dotnet/runtime/issues/53018", TestPlatforms.Browser)]
+                return;
+            }
```
Comment (radical) [inline src/libraries/Common/tests/System/Net/Http/PostScenarioTest.cs:90]: Doesn't the `ActiveIssue` disable this for wasm completely? why is this needed?

File: src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTestBase.SocketsHttpHandler.cs (hunk 12 above)
Comment (pavelsavara) [reply src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTestBase.SocketsHttpHandler.cs:19]: https://github.com/dotnet/runtime/commit/33d881fd04242635e3d667f93acaad433baa308f#diff-1d4ce147edca0ce81f312b96fce76ac52eb02eb6d04e660817e3b47230edc794R9

File: src/libraries/Common/tests/System/Net/Http/PostScenarioTest.cs (hunk 13 above)
Comment (pavelsavara) [reply src/libraries/Common/tests/System/Net/Http/PostScenarioTest.cs]: you are right, fixed

File: src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTestBase.SocketsHttpHandler.cs (hunk 12 above)
Comment (radical) [reply src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTestBase.SocketsHttpHandler.cs:19]: IIUC, `QuicImplementationProviders.MsQuic` will throw PNSE. And the supported platforms are at least partially being encoded via the target frameworks in the project file. So, IMHO, instead of explicitly adding `IsNotBrowser` here, I think we should return `QuicImplementationProviders.MsQuic.IsSupported`, and if it throws PNSE, then return `false`.

Comment (pavelsavara) [reply src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTestBase.SocketsHttpHandler.cs]: Okay

**Review by radical — APPROVED** [review]
Thanks for addressing the feedback. Assuming everything passes, LGTM, thanks 👍
